File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> ans = new LinkedList<>();
        if(root == null) return ans;
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        boolean order = true;
        while(!q.isEmpty())
        {
            int levelsize = q.size();
            List<Integer> temp = new LinkedList<>();
            for(int i=0; i<levelsize; i++)
            {
                TreeNode curr = q.poll();
                temp.add(curr.val);
                if(curr.left != null)
                    q.add(curr.left); 
                if(curr.right != null)
                    q.add(curr.right);  
            }
            if(!order)
                Collections.reverse(temp);
            ans.add(temp);
            order = !order;
        }
        return ans;
    }
}

File: 0103-binary-tree-zigzag-level-order-traversal/0103-binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> ans;
        if(!root) return ans;
        queue<TreeNode*> q;
        q.push(root);
        bool order = true;
        while(!q.empty())
        {
            int levelsize = q.size();
            vector<int> temp;
            for(int i=0; i<levelsize; i++)
            {
                TreeNode*curr = q.front();
                q.pop();
                temp.push_back(curr->val);
                if(curr->left)
                    q.push(curr->left); 
                if(curr->right)
                    q.push(curr->right);  
            }
            if(!order)
                reverse(temp.begin(), temp.end());
            ans.push_back(temp);
            order = !order;
        }
        return ans;
    }
};